### PR TITLE
Multi catalogs

### PR DESCRIPTION
**Pull request by zachidan**

### CODE REVIEW

File: store/catalog_map.go [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+	"sync"
+
+	log "github.com/Sirupsen/logrus"
+
+	"github.com/amalgam8/registry/auth"
+	"github.com/amalgam8/registry/utils/logging"
+)
+
+const (
+	module string = "STORE"
```
Comment (zcahana) [inline store/catalog_map.go]: The explicit `string` is redundant.

File: store/catalog_map.go [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+	defer cm.Unlock()
+
+	catalog, exists := cm.catalogs[namespace]
+	if exists {
+		return catalog, nil
+	}
+
+	catalog, err := cm.catalogFactory.CreateCatalog(namespace)
+	if err != nil {
+		return nil, err
```
Comment (zcahana) [inline store/catalog_map.go]: Should probably log the error.

File: store/config.go [hunk 3]
```diff
@@ -38,17 +40,22 @@ type Config struct {
 	NamespaceCapacity int
 
 	SyncWaitTime time.Duration
+
+	AddsOn      []CatalogFactory
```
Comment (zcahana) [inline store/config.go]: `Addons` typo?
Anyhow, `AddonCatalogs` or `ExtensionCatalogs` could be a better fit.

File: store/inmem_catalog.go [hunk 4]
```diff
@@ -45,12 +44,34 @@ const (
...
+
+type inmemConfig struct {
+	defaultTTL time.Duration
+	minimumTTL time.Duration
+	maximumTTL time.Duration
+
+	namespaceCapacity int
+}
+
+type inmemFactory struct {
```
Comment (zcahana) [inline store/inmem_catalog.go]: This file declares 3 types: `inmemConfig`, `inMemoryCatalog`, and `inMemoryService`.
For consistency, I would rename all three to either `inmemXXX` or `inMemoryXXX`.

File: store/multi_catalog.go [hunk 5]
```diff
@@ -72,13 +72,43 @@ func (mc *multiCatalog) SetStatus(instanceID, status string) error {
 }
 
 func (mc *multiCatalog) List(serviceName string, predicate Predicate) ([]*ServiceInstance, error) {
-	return mc.catalogs[0].List(serviceName, predicate)
+	instanceCollection := make([]*ServiceInstance, 0, 10)
+	for _, catalog := range mc.catalogs {
+		list, err := catalog.List(serviceName, predicate)
+		if err == nil {
```
Comment (zcahana) [inline store/multi_catalog.go]: In case of an error - should log it, as it doesn't even propagate up...

File: store/multi_catalog.go [hunk 6]
```diff
@@ -72,13 +72,43 @@ func (mc *multiCatalog) SetStatus(instanceID, status string) error {
...
+	}
+
+	if len(instanceCollection) == 0 {
+		return nil, NewError(ErrorNoSuchServiceName, "no such service ", serviceName)
+	}
+
+	return instanceCollection, nil
 }
 
 func (mc *multiCatalog) Instance(instanceID string) (*ServiceInstance, error) {
```
Comment (zcahana) [inline store/multi_catalog.go]: Shouldn't `Instance()` (being a read-only operation) iterate through the catalogs until it finds the instance?

File: store/multi_catalog.go [hunk 7]
```diff
@@ -72,13 +72,43 @@ func (mc *multiCatalog) SetStatus(instanceID, status string) error {
...
+		}
+	}
+
+	services := make([]*Service, 0, len(smap))
+	for _, svc := range smap {
+		services = append(services, svc)
+	}
+
+	return services
 }
```
Comment (zcahana) [inline store/multi_catalog.go]: So, `multiCatalog.catalogs[0]` seems to be a special catalog - the only one that is read-write.
Shouldn't this be documented somehow? Either in the documentation of multiCatalog and/or its config object, or maybe as separate configuration fields (RWCatalog, ROCatalogs)

File: config/flags.go [hunk 8]
```diff
@@ -45,6 +45,8 @@ const (
 	DefaultTTLFlag        = "default_ttl"
 	MaxTTLFlag            = "max_ttl"
 	MinTTLFlag            = "min_ttl"
+
+	K8sURLFlag = "k8s_url"
```
Comment (zcahana) [inline config/flags.go]: Need to update the registry's README.me with this flag and its usage.

File: store/kubernetes/catalog.go [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+func newK8sCatalog(namespace auth.Namespace, client *k8sClient) (*k8sCatalog, error) {
+	catalog := &k8sCatalog{
+		services:  serviceMap{},
+		instances: instanceMap{},
+		namespace: namespace,
+		client:    client,
+		logger:    logging.GetLogger(module),
+	}
+
+	catalog.refresh()
```
Comment (zcahana) [inline store/kubernetes/catalog.go]: Refreshing synchronously here is potentially unwanted:
Consider a completely new tenant (namespace), doing the first registration call (POST /api/v1/instances).
This operation creates a multi-catalog for him, which in turn creates an in-memory-catalog, replicated-catalog, k8s-catalog, ..... and synchronously refreshes with k8s API.

Alternative: `go catalog.refresh()` in here too, but when `ListServices()`, `ListInstances()` or `Instance()` are called, check if no refresh has already been done, and if needed - force refresh.

File: store/kubernetes/catalog.go [hunk 10]
```diff
@@ -176,7 +187,7 @@ func (kc *k8sCatalog) getServices() (serviceMap, instanceMap, error) {
 						ServiceName: sname,
 						Endpoint:    &store.Endpoint{Type: endpointType, Value: endpointValue},
 						Status:      "UP",
-						Metadata:    []byte(fmt.Sprintf("{\"kubernetes_url\":\"%s/%s\"}", kc.client.getEndpointsURL(kc.namespace), sname)),
+						Metadata:    []byte(fmt.Sprintf("{\"kubernetes_url\":\"%s/%s\", \"version\":\"%s\"}", kc.client.getEndpointsURL(kc.namespace), sname, version)),
```
Comment (zcahana) [inline store/kubernetes/catalog.go]: A more friendly/readable way can be to just marshal a map:
```go
metadataBytes, _ := json.Marshal(map[string]interface{}{
	"kubernetes_url": fmt.Sprintf("%s/%s", kc.client.getEndpointsURL(kc.namespace), sname),
	"version": version,
})
```

File: store/kubernetes/client.go [hunk 11]
```diff
@@ -50,6 +51,11 @@ func newK8sClient(k8sURL string) (*k8sClient, error) {
 		return nil, err
 	}
 
+	t, err := ioutil.ReadFile("/var/run/secrets/kubernetes.io/serviceaccount/token")
```
Comment (zcahana) [inline store/kubernetes/client.go]: I'm guessing this file gets mounted into k8s containers by the k8s daemon?
What happens if we want to run the a k8s-enabled registry on a non-k8s node?
We should probably switch to a more general mechanism of hooking up with the k8s API token.

File: store/kubernetes/client.go [hunk 12]
```diff
@@ -74,7 +81,10 @@ func (client *k8sClient) getEndpointsURL(namespace auth.Namespace) string {
 func (client *k8sClient) getEndpointsList(namespace auth.Namespace) (*EndpointsList, error) {
 	endpointsList := EndpointsList{}
 
-	resp, err := client.httpClient.Get(client.getEndpointsURL(namespace))
+	req, _ := http.NewRequest("GET", client.getEndpointsURL(namespace), nil)
+	req.Header.Set("Authorization", client.k8sAuth)
+	resp, err := client.httpClient.Do(req)
+	//resp, err := client.httpClient.Get(client.getEndpointsURL(namespace))
```
Comment (zcahana) [inline store/kubernetes/client.go]: The commented-out line can definitely be cleaned-out.

**Comment by zcahana** [comment]
@zachidan @frankbu see comments posted above.

File: store/kubernetes/client.go (hunk 11 above)
Comment (frankbu) [inline store/kubernetes/client.go]: @zcahana Yes this only works if the registry is running in k8s itself. I agree we should figure out a general way to get the token. I just didn't have time to figure out how, and thought that this is sufficient for now.

File: store/kubernetes/client.go [hunk 13]
```diff
@@ -82,9 +91,10 @@ func (client *k8sClient) getEndpointsList(namespace auth.Namespace) (*EndpointsL
 	endpointsList := EndpointsList{}
 
 	req, _ := http.NewRequest("GET", client.getEndpointsURL(namespace), nil)
-	req.Header.Set("Authorization", client.k8sAuth)
+	if client.k8sToken != "" {
+		req.Header.Set("Authorization", fmt.Sprintf("bearer %s", client.k8sToken))
```
Comment (zcahana) [inline store/kubernetes/client.go]: According to [this](http://kubernetes.io/docs/admin/authentication/), K8s API server expectes a _Bearer_, and not a _bearer_ prefix. I'm not sure what has been implemented in there, but we better stick to _Bearer_.

**Comment by zcahana** [comment]
@zachidan, see a single comment above (_Bearer_ token), but otherwise LGTM.

**Comment by zachidan** [comment]
@zcahana Thanks a lot for your helpful review. I'm done with your review.

**Comment by zcahana** [comment]
@zachidan can this PR be merged from your perspective?

**Comment by zachidan** [comment]
@zcahana Yes. This PR is ready to be merged.

**Pull request merged**